x-pack/plugins/apm/server/lib/services/get_service_details.ts

```diff
@@ -0,0 +1,182 @@
+import { Setup, SetupTimeRange } from '../helpers/setup_request';
+
+type ServiceDetails = Pick<
+  TransactionRaw,
+  'service' | 'agent' | 'host' | 'container' | 'kubernetes' | 'cloud'
+>;
+
+interface ServiceDetailsResponse {
+  service?: {
+    versions?: string[];
+    runtime?: {
+      name: string;
+      version: string;
+    };
+    framework?: string;
+    agent: {
+      name: string;
+      version: string;
+    };
+  };
+  container?: {
+    os?: string;
+    isContainerized?: boolean;
+    avgNumberInstances?: number;
+    orchestration?: 'Kubernetes' | 'Docker';
+  };
+  cloud?: {
+    provider?: string;
+    availabilityZones?: string[];
+    machineTypes?: string[];
+    projectName?: string;
+  };
+}
+
+export async function getServiceDetails({
+  serviceName,
+  setup,
+}: {
+  serviceName: string;
+  setup: Setup & SetupTimeRange;
+}): Promise<ServiceDetailsResponse> {
+  const { start, end, apmEventClient } = setup;
+  const { intervalString } = getBucketSize({ start, end });
+
+  const filter = [
+    { term: { [SERVICE_NAME]: serviceName } },
+    { term: { [PROCESSOR_EVENT]: ProcessorEvent.transaction } },
+    { range: rangeFilter(start, end) },
+    ...setup.esFilter,
+  ];
+
+  const should = [
+    CONTAINER,
+    KUBERNETES,
+    CLOUD,
+    HOST,
+    AGENT,
+  ].map((fieldName) => ({ exists: { field: fieldName } }));
+
+  const params = {
+    apm: {
+      events: [ProcessorEvent.transaction],
+    },
+    body: {
+      size: 1,
+      _source: [SERVICE, AGENT, HOST, CONTAINER, KUBERNETES, CLOUD],
+      query: { bool: { filter, should } },
+      aggs: {
+        serviceVersions: {
+          terms: {
+            field: SERVICE_VERSION,
+            size: 10,
+            order: { _key: 'desc' },
+          },
+        },
+        availabilityZones: {
+          terms: {
+            field: CLOUD_AVAILABILITY_ZONE,
+            size: 10,
+          },
+        },
+        machineTypes: {
+          terms: {
+            field: CLOUD_MACHINE_TYPE,
+            size: 10,
+          },
+        },
+        histogram: {
+          date_histogram: {
+            field: '@timestamp',
+            fixed_interval: intervalString,
+            min_doc_count: 0,
+            extended_bounds: { min: start, max: end },
+          },
+          aggs: {
+            total_instances: { cardinality: { field: SERVICE_NODE_NAME } },
+          },
+        },
+        avgNumberInstances: {
+          avg_bucket: { buckets_path: 'histogram>total_instances' },
+        },
+      },
+    },
+  };
+
+  const response = await apmEventClient.search(params);
+
+  if (response.hits.total.value === 0) {
+    return {
+      service: undefined,
+      container: undefined,
+      cloud: undefined,
+    };
+  }
+
+  const { service, agent, host, kubernetes, container, cloud } = response.hits
+    .hits[0]._source as ServiceDetails;
+
+  const serviceDetails: ServiceDetailsResponse['service'] = {
```

Comment:
Is this type necessary? I thought the ES types were smart enough to infer this stuff